src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,322 @@
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, &trajVel);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, &trajAccel);
+
+    if ( trajPos.size() > 0 )
+      reorder(indexMap, trajPos, _positions);
+    if ( trajVel.size() > 0 )
+      reorder(indexMap, trajVel, _velocities);
+    if ( trajAccel.size() > 0 )
+      reorder(indexMap, trajAccel, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+//=============================================================================
+// The rows of inVector is reordered in outVector.
+void reorder(std::map<size_t, size_t> indexMap,
+  const Eigen::VectorXd& inVector, Eigen::VectorXd* outVector)
+{
+  *outVector = Eigen::VectorXd::Zero(inVector.size());
+  for (auto it = indexMap.begin(); it != indexMap.end(); ++it)
+  {
+    (*outVector)(it->second) = inVector(it->first);
+  }
+}
+
+//=============================================================================
+std::vector<const dart::dynamics::Joint*> findJointByName(
+  const dart::dynamics::MetaSkeleton& metaSkeleton,
+  const std::string& jointName)
+{
+  std::vector<const dart::dynamics::Joint*> joints;
+
+  for (size_t i = 0; i < metaSkeleton.getNumJoints(); ++i)
+  {
+    if (metaSkeleton.getJoint(i)->getName() == jointName)
+    {
+      joints.emplace_back(metaSkeleton.getJoint(i));
+    }
+  }
+
+  return joints;
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const trajectory_msgs::JointTrajectory& jointTrajectory)
+{
+  if (!space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = space->getNumSubspaces();
+  if (jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that the names in jointTrajectory are unique.
+  std::vector<std::string> joint_names;
+  joint_names.reserve(numControlledJoints);
+  for (size_t i = 0; i < numControlledJoints; ++i)
+  {
+    joint_names.emplace_back(jointTrajectory.joint_names[i]);
+  }
+  std::sort(joint_names.begin(), joint_names.end());
+  for (size_t i = 0; i < numControlledJoints - 1; ++i)
+  {
+    if (joint_names[i] == joint_names[i+1])
+    {
+      std::stringstream message;
+      message << "JointTrajectory has multiple joints with same name ["
+        << joint_names[i] << "].";
+      throw std::invalid_argument{message.str()};
+    }
+  }
```

Comment:
Nit: it would be cleaner to replace this loop with:
```c++
auto duplicate_it = std:: adjacent_find(std::begin(joint_names), std::end(joint_names));
if (duplicate_it != std::end(joint_names)) { /* ... */ }
```